### DISCUSSION TITLE
Using mqtt_as and trying to publish just before client close called?

### DISCUSSION THREAD

**Discussion by Jibun-no-Kage** [micropython/micropython#11264] [General]
I have an ESP32 based micropython script that publishes as it starts and as it stops.  Started after boot, and Stopped as Keyboard interrupt or script exit due to error, i.e. finally section publish.

The startup publish works fine.  I setup the client, do subscribes, then publish the script is listening for messages.  However, the not listening or script is ending, publish never gets to the broker.  On restart of E$P32 I do get the 'will' publish from the broker.  I have tried a couple of things.

1) publish via MQTT client with out using asyncio methods, thinking the main thread would handle it, but the main is really under asyncio.run control
2) adding asyncio sleep, thinking the MQTT client is closing to fast to let any pending messages get the broker

In traditional python I did this via a thread and queue, and thread does not exit until the queue is empty.  But using asyncio, methods, not working.  Each publish is invoked via create_task asyncio method.  So I thought using an await asychio,sleep sufficiently long delay would let the in progress publish task complete?  But apparently not so?

**Comment by peterhinch** [comment]
I don't think this is possible using `mqtt_as` because the `uasyncio` scheduler will have been stopped by the keyboard int or exception. This is the purpose of the last will feature of MQTT: the broker publishes a message when the client has crashed.

To do this I think you'll need to use synchronous code and adapt the official synchronous driver.

**Comment by Jibun-no-Kage** [comment]
Not surprised by your comment, I think it was what I was expecting to hear.  I can live with the will message as notification of shutdown/reboot on MQTT subscriber side.  It would have been nice to communicate a bit more actively.
